proposals/p3763.md

```diff
@@ -0,0 +1,858 @@
+    in different libraries, or even different packages, so such coupling would
+    make refactorings that change the way that code is expressed but not its
+    meaning either difficult or impossible.
+-   The associated function in an `impl` is expected to have different syntax
+    than that in the interface in some cases. The two declarations are in
+    different scopes, so will refer to the same types in different ways. And the
+    declaration in the `impl` is declared with knowledge of the `Self` type and
+    associated constants for the interface, which it may be reasonable to use
+    directly in the declaration of the function, as in the preceding example.
+
+Therefore, different rules are used. In the specific case where an associated
+function declaration in the `impl` can be used directly to satisfy a requirement
+introduced by a function declaration in the `interface`, it is used directly.
+For example, this is necessary to avoid infinite recursion when implementing the
+`Call` interface.
+
+The function in the `impl` is used directly when:
+
+-   Each parameter in the `impl` function has the same type as the parameter in
+    the `interface`. This includes the `self` parameter, which must be present
+    in both functions if it is present in either.
+-   Each parameter in the `impl` has the same category -- either `var` or not --
+    as the parameter in the `interface`.
+-   The return type in the `interface` and `impl` are the same type.
+
+> **Note:** More constraints are expected to appear here over time. The key
+> property we aim to identify is whether the two functions have the same calling
+> convention.
+
+Otherwise, a synthetic function called a _thunk_ is generated:
+
+-   The declaration of the thunk is formed by substituting the `Self` type of
+    the `impl` into the declaration in the `interface`. This implicitly also
+    provides values for any associated constants used in the declaration.
+-   The body of the thunk calls the function in the `impl`, passing in the
+    arguments to the thunk, and, if a return type is specified in the interface,
+    returning the value returned by the call.
+
+If the function in the interface does not have a return type, the program is
+invalid if the function in the `impl` specifies a return type.
+
+> **Note:** Another rule might work better here. For example, we could allow
+> `impl`s to add a return type in anticipation of the `interface` later adding
+> one. However, such a feature is considered out of scope for this proposal.
+
+Implicit conversions are performed as necessary to initialize the parameters of
+the function in the `impl` from the parameters of the thunk, and to initialize
+the return value of the function from the result of the call.
+
+It is an error if a thunk is needed to wrap a function declaration with a `var`
+parameter, because otherwise a copy would always be performed when initializing
+the parameter.
```

Comment:
Should we add a note about potentially revisiting this when we have moves?

Something like:

```suggestion
It is an error if a thunk is needed to wrap a function declaration with a `var`
parameter, because otherwise a copy would always be performed when initializing
the parameter.

> **Note:** Here as well another rule may work better and we should revisit in
> the future. Specifically, it might be better to allow these cases and simply
> move from the outer `var` to the inner `var` even though this still leaves
> two allocations in principle.
```